packages/reward-root-submitter/tests/test_submitter.py

```diff
@@ -0,0 +1,267 @@
+        "tests/resources/reward_output",
+    )
+    submitter.submit_root(
+        reward_programs()[0],
+        12,
+        "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+    )
+    hw = reward_pool.caller.payeeRoots(reward_programs()[0], 12)
+    assert hw == HexBytes(
+        "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+    )
+
+
+def test_submit_root_twice_is_safe(w3, reward_pool, deploy_address, deploy_key):
+    submitter = RootSubmitter(
+        w3,
+        deploy_address,
+        deploy_key,
+        reward_pool.address,
+        "tests/resources/reward_output",
+    )
+    submitter.submit_root(
+        reward_programs()[0],
+        12,
+        "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+    )
+    hw = reward_pool.caller.payeeRoots(reward_programs()[0], 12)
+    assert hw == HexBytes(
+        "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+    )
+    # Now try to submit again
+    submitter.submit_root(
+        reward_programs()[0],
+        12,
+        "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+    )
+    hw = reward_pool.caller.payeeRoots(reward_programs()[0], 12)
+    assert hw == HexBytes(
+        "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+    )
+
+
+def test_submits_all_roots(w3, reward_pool, deploy_address, deploy_key):
+    submitter = RootSubmitter(
+        w3,
+        deploy_address,
+        deploy_key,
+        reward_pool.address,
+        "tests/resources/reward_output",
+    )
+    submitter.submit_all_roots()
+    # There is no root for  "0xA2e8225dE0385ebC20B3C0160864f4e20a750cfd" / 24098304
+    for reward_program_id in reward_programs():
+        for payment_cycle in [24000000, 24065536]:
+            assert reward_pool.caller.payeeRoots(
+                reward_program_id, payment_cycle
+            ) != HexBytes(
+                "0x0000000000000000000000000000000000000000000000000000000000000000"
+            )
+
+
+def test_can_find_all_payment_cycles():
+    all_cycles = list(get_all_reward_outputs(AnyPath("tests/resources/reward_output/")))
+    assert len(all_cycles) == 7
+    assert {
+        "file": AnyPath(
+            "tests/resources/reward_output/rewardProgramID=0xA2e8225dE0385ebC20B3C0160864f4e20a750cfd/paymentCycle=24000000/results.parquet"
+        ),
+        "reward_program_id": "0xA2e8225dE0385ebC20B3C0160864f4e20a750cfd",
+        "payment_cycle": 24000000,
+    } in all_cycles
+
+
+def test_can_find_all_payment_cycles_on_s3():
+    s3_client = LocalS3Client(local_storage_dir="tests")
+    all_cycles = list(
+        get_all_reward_outputs(s3_client.S3Path("s3://resources/reward_output/"))
+    )
+    assert len(all_cycles) == 7
+    assert {
+        "file": s3_client.S3Path(
+            "s3://resources/reward_output/rewardProgramID=0xA2e8225dE0385ebC20B3C0160864f4e20a750cfd/paymentCycle=24000000/results.parquet"
```

Comment:
Is this s3:bucket have privileged permission to this submitted?

Reply:
This should be a faked s3 bucket, the LocalS3Client pretends to be s3 for users but just maps to a folder locally. This ended up helping for the tests as I originally had globbing which worked on a filesystem but broke on s3

Reply:
Ahh ok. I came across this (it getting github action to work). Thanks it was useful.